scrapy/core/spidermw.py

```diff
@@ -40,36 +44,73 @@ def process_spider_input(response):
             for method in self.methods['process_spider_input']:
                 try:
                     result = method(response=response, spider=spider)
-                    assert result is None, \
-                            'Middleware %s must returns None or ' \
-                            'raise an exception, got %s ' \
-                            % (fname(method), type(result))
+                    if result is not None:
+                        raise _InvalidOutput('Middleware {} must return None or raise an exception, got {}' \
+                                             .format(fname(method), type(result)))
+                except _InvalidOutput:
+                    raise
```

Comment:
Just to check that I'm understanding the code correctly, why do we re-raise ``_InvalidOutput`` here?

Reply:
Ah never mind, I see why - to avoid it falling into the next ``except`` clause.